### PR TITLE
Fix #1910. Add support for site aliases that point into a Docker container

### LINKED ISSUE #1910: Feature request: Add docker exec to aliases remote
Currently it is able to execute remote drush by alias like this:

```
$aliases['stage'] = array(
    'remote-host' => 'mystagingserver.myisp.com',
    'remote-user' => 'publisher',
);
```

But today we run Drupal also in Docker container, so it is good to run Drush commands also inside it. It could be configured like this:

```
$aliases['stage'] = array(
    'docker-machine' => 'aws01',
    'docker-container' => 'drupal_web_1',
    'docker-user' => 'www-data',
);
```

And also conbined to connect to server and there into container:

```
$aliases['stage'] = array(
    'remote-host' => 'mystagingserver.myisp.com',
    'remote-user' => 'docker',
    'docker-container' => 'drupal_web_1',
    'docker-user' => 'www-data',
);
```

So we will be able to use it intuitively like this:

```
drush sql-sync @site.local @site.stage
```

Workaround (how we have to to it today to sync into container):

```
drush @site.dev sql-dump | docker exec -i —user=www-data drupal_web_1 drush sql-cli
```

### PR DESCRIPTION
**Pull request by weitzman**

### CODE REVIEW

**Comment by greg-1-anderson** [comment]
Could we perhaps provide this functionality in the Site Process library, and not have to add special checking in Drush?

**Comment by greg-1-anderson** [comment]
Could we go through the exercise of defining what we would have to do to make sql:sync work with a docker-based Drupal site? What command would we use to dump the mysql database? Would we be able to do this in such a way that the target dump is written to a shared path that is readable by the local process (i.e. rsync step not needed)? What options would we need to pass to `docker-compose exec` to make that work? How would the sql:sync command know what to do vis-a-vis the rysnc portion of the command? What properties would we have to define in the site alias?

File: sut/drush/sites/sut.site.yml [hunk 1]
```diff
@@ -22,3 +22,9 @@ stage:
   uri: stage
 prod:
   uri: prod
+
+# An example of a Docker Compose environment. Used for ad hoc testing, and not by Unish.
+local:
+  service: drupal
```
[the remark this answers is not in the document]
Comment (weitzman) [reply sut/drush/sites/sut.site.yml]: Correct. In this case I refer to our own [docker-compose.yml](https://github.com/drush-ops/drush/blob/master/docker-compose.yml).

**Comment by weitzman** [comment]
sql:sync is a good question. Assuming that rsync is out of the question, then we need to rely on Docker's array of [persistence options](https://thenewstack.io/methods-dealing-container-storage/). In general, I think folks will want to setup a [Named Volume](https://docs.docker.com/engine/reference/commandline/volume_create/#/driver-specific-options) and write/read the dump file from there. sql:sync already has a `--no-sync` and related options which should be enough.

I may enhance our docker-compose.yml with a `drupal2` service just as a POC for Docker based sql:sync.

There was some discussion in #1906 about this but no-one said how sql:sync is supposed to work across docker powered environments.

**Comment by greg-1-anderson** [comment]
Yeah, getting really close. We just need to figure out what the correct "remote" vs "not remote" determinations should be. Currently the notion of "remote" is overloaded. We need to split it up so that the key options that command line you made can be injected into the docker site alias.

**Comment by sjerdo** [comment]
What does issue #1906 have to do with this PR? I'm assuming it should be #1910.

I don't think we could assume a sync always is between two docker containers when one of the aliases uses docker for transport. And even if so, they won't share a volume just for syncing files.
Rsync could be used to sync files (or a sql dump) between containers and/or aliases (Although the local host should be temporarily used to store the file(s)). 
E.g. We're currently syncing files using rsync for one of our clients using the following command using kubernetes for the remote-shell command in rsync (-e or --rsh parameter)
`rsync -akz --blocking-io --rsh="kubectl exec -i project-pod-id" --delete --exclude="excluded.file.php" -- "dir/" --:"/var/www/path/to/external/dir"` or 
`rsync -akz --blocking-io --rsh="kubectl exec -i" --delete --exclude="excluded.file.php" dir/ project-pod-id:/var/www/path/to/external/dir`
The same could be used for rsyncing files to or from a docker container:
`rsync -akz --rsh="docker exec -i" --delete --exclude="excluded.file.php" dir/ container:"/var/www/path/to/external/dir"` or 
`rsync -akz --rsh="docker exec -i" --delete container:"/var/www/path/to/external/file.txt" file.txt`

**Comment by greg-1-anderson** [comment]
We're not going to solve the sql:sync problem in this PR, but comments on the different permutations on how to handle this are appreciated. Drupal-on-Kube is another interesting use-case for sure.

File: sut/drush/sites/sut.site.yml [hunk 2]
```diff
@@ -23,21 +23,13 @@ stage:
 prod:
   uri: prod
 
-# An example of a Docker Compose environment. Used for ad hoc testing, and not by Unish.
+# An example of a Docker Compose environment. See exa,ple.site.yml for a bit more info.
```
Comment (greg-1-anderson) [inline sut/drush/sites/sut.site.yml]: typo: exa,ple.site.yml

**Comment by weitzman** [comment]
@sjerdo thanks for those rsync examples. I had not seen that form before. Perhaps you could openb a PR with new Examples for the drush rsync command. Thats the best place I can think of to document this.

**Comment by greg-1-anderson** [comment]
🎉

**Comment by leymannx** [comment]
Everybody coming here via search engine or from the other issue, you now can use the following snippet in your `self.site.yml` to execute Drush commands in a remote container with `drush @dev cr`.

```yaml
dev:
  host: my.remote.host.example.com
  user: myuser
  docker:
    service: my-container-name
```

**Comment by leymannx** [comment]
Sorry, last comment was probably based on a false conclusion. We had a Drush on that host itself as well as on that host in a container. I'm not sure if it's supposed to be working like that at all (as in SSH into host and execute command in a container on that host).

**Comment by leymannx** [comment]
@weitzman – Do I understand this right, the Docker support is not meant to be used to connect to a remote host that's running Docker containers, and to exec Drush then inside this remote's host containers? It's meant to be used to execute Drush commands in a respectively local Docker container?

**Comment by weitzman** [comment]
@leymannx - my understanding is that your scenario works fine. This is an abstraction provided by Docker itself. Just set DOCKER_HOST env variable to point to the remote docker.

**Comment by csheinin** [comment]
Soo.... here's what I'm deriving from this for my <sitename>.site.yml:
`local:
  docker:
    service: drupal
    exec:
      options: --user root
  uri: dev
  root: /var/www/html/web
  paths:
    drush-script: /var/www/html/drush`
  
  Does this imply we should make sure we have drush installed inside the docker instance itself? Does anyone have time to jot down some notes of "what is required so that drush pm-list command from my local box returns output from target drupal-in-container"?

**Comment by Trogie-Vynckier** [comment]
Nice! Any idea how to combine ssh to a machine that runs drupal/drush in a container?

**Comment by lukasfischer** [comment]
As I understand, th use case ssh to remote + entering docker container on remote server is not supported. I created the following bash script - it assumes you use drupal with drush and lando:

`
ssh -t $DOCKER_PROD_USER@$DOCKER_PROD_HOST "cd $DOCKER_PROD_DOCKER_PATH; exec docker compose exec -T drupal sh -c 'drush sql-dump --structure-tables-list=cache,cache_*,search_index,watchdog'" > dump.sql
lando db-import dump.sql
rm dump.sql
`

If there is a more elegant solution using drush site alias, I'm curious.

fyi @Trogie-Vynckier

**Pull request merged**